design/priority-class-name-support_design.md

```diff
@@ -0,0 +1,332 @@
+
+2. **Using a configuration file**: We could have added support for specifying the priority class names in a configuration file. However, this would have required additional changes to the Velero CLI and would have been more complex to implement.
+
+3. **Inheriting priority class from parent components**: We initially considered having maintenance jobs inherit their priority class from the Velero server, and data movers inherit from the node agent. However, this approach doesn't allow for the appropriate prioritization of different components based on their importance and resource requirements.
+
+## Security Considerations
+There are no security considerations for this feature.
+
+## Compatibility
+This feature is compatible with all Kubernetes versions that support priority classes. The PodPriority feature became stable in Kubernetes 1.14. For more information, see the [Kubernetes documentation on Pod Priority and Preemption](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/).
+
+## Implementation
+The implementation will involve the following steps:
+1. Add the priority class name fields to the `VeleroOptions` struct
+2. Add the priority class name fields to the `podTemplateConfig` struct
+3. Add the `WithPriorityClassName` functions for different components
+4. Update the `Deployment` function to use the server priority class name
+5. Update the `DaemonSet` function to use the node agent priority class name
+6. Add the necessary functions to the `veleroutil` package for retrieving priority class names
+7. Update the maintenance job creation to use the maintenance job priority class name
+8. Update the data mover pod creation to use the data mover priority class name
+9. Add the priority class name flags to the `velero install` command
+
+Note: We're only adding the flags to the CLI install command, not to the server configuration, as the priority class names are only needed during installation. The priority class names will be stored in the respective Kubernetes resources during installation.
+
+This approach ensures that different Velero components can use different priority class names based on their importance and resource requirements:
+
+1. The Velero server deployment can use a higher priority class to ensure it continues running even under resource pressure.
+2. The node agent daemonset can use a medium priority class.
+3. Maintenance jobs can use a lower priority class since they should not run when resources are limited.
+4. Data mover pods can use a lower priority class since they should not run when resources are limited.
+
+### Implementation Considerations
+
+For maintenance jobs and data movers, we need to ensure that the priority class names are stored in a way that they can be accessed when these components are created, which may happen after installation. There are several approaches we could take:
```

Comment:
For fs-backup, considering https://github.com/vmware-tanzu/velero/issues/8809, we also need to consider the configuration for pvb/pvr pods, that should be the same with data mover, in node-agent-configmap